### DISCUSSION TITLE
Metadata portability?  ABS vs OPF vs NFO vs embed, etc

### DISCUSSION THREAD

**Discussion by jeff47** [advplyr/audiobookshelf#1806] [General]
I would like to make sure my metadata is portable across platforms, or in case something happens to my ABS database.  I know that ABS will let us save ABS files in each directory -- **(1)** Is that instead of in the main database, or in addition to?

I saw an [issues thread](https://github.com/advplyr/audiobookshelf/issues/141) opened up to save as OPF and/or NFO, and the various issues involved with either.  It appears that the resolution was to save as ABS instead -- is that correct?

I'm not sure what the best approach is, but after so much work getting everything matched, I want to make sure it doesn't get lost.

**Comment by advplyr** [comment]
I'm not sure if you saw but on the last release I added the option to save as `JSON` instead. This is much better and will be the default for all new installs. The `.abs` file had issues dealing with commas and wasn't friendly for other software to use.

**Reply by jeff47** [reply]
Oh, just saw that option - that looks good.  I didn't notice the new option on that page.

How do I generate for JSON files for all existing books?

If I have `Store metadata with item` turned on, does that mean things are *only* stored in the book directory, or is that in addition to the `metadata/items`?

**Reply by advplyr** [reply]
If you change to JSON then just do a regular scan it should replace them all right away. That will only happen though if they are stored in your book directory. If you change to JSON and they are in `metadata/items` then they will update to JSON the first time the book is updated.

If you use the Store metadata with item setting it will only store it in your book directory.

I've been thinking though of changing this so it stored in both places. Also, the `metadata/items` is read-only in the sense that if you change the metadata file Abs won't pick up your changes. If you are storing it in the book directory though Abs will use it as a source of truth for your metadata and updates will be picked up.

It's a little weird and confusing I know. I would like to make this more intuitive and document it but I'm not sold on one way or the other.

**Reply by jeff47** [reply]
> I've been thinking though of changing this so it stored in both places. Also, the `metadata/items` is read-only in the sense that if you change the metadata file Abs won't pick up your changes. If you are storing it in the book directory though Abs will use it as a source of truth for your metadata and updates will be picked up.
> 
> It's a little weird and confusing I know. I would like to make this more intuitive and document it but I'm not sold on one way or the other.

I can see the problem.  You'd want to make sure they don't get out of sync, and figure out which one "wins" if they do.  I'm not sure it needs to be in both places unless there's something intrinsic to how the server/player operates.  At the moment, I slightly prefer having it stored with the audio files, but I'm not sure I have a good reason for that.

What happens if people aren't using a single book per directory?  Might run into issues there with filename collisions and it not being clear what metadata belongs to which file.  Something to consider, if people don't follow the recommended layout.  I've been tempted when I have a single file M4B to skip the author/title directory format.

**Reply by advplyr** [reply]
Yeah for a long time it was required that each book be in its own directory. When I added the option to store single file audiobooks in the root folder I had to add an extra check to always use `metadata/items` for those.
